### ISSUE TITLE
[patch] Persistent transactions for adbapi

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6830]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| grib reported|
|-|-|
|Trac ID|trac#303|
|Type|enhancement|
|Created|2003-10-01 01:22:00Z|

Attachments:

* [diff.out](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbc/bbcbb1e844266f4abdfc29b3d8a64628607fa47e/47f540b04f7bd7f98ecbdfc50c459e62638e70e0.out) (4603 bytes) - added by grib on 2003-10-01 01:22:00Z - 
* [adbapi.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbc/bbcbb1e844266f4abdfc29b3d8a64628607fa47e/1f77fc1e67eb02de7f7cc7ec0c3634079150b9ab.diff) (4651 bytes) - added by grib on 2003-10-04 19:37:07Z - 
* [adbapi.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bbc/bbcbb1e844266f4abdfc29b3d8a64628607fa47e/17b0a53af028aba27c2534c0321a7bfea9195289.diff) (5119 bytes) - added by grib on 2003-10-07 21:42:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__303 303
type__enhancement enhancement
reporter__grib grib
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__core core
keywords__ 
time__1064971320000000 1064971320000000
changetime__1468230992077369 1468230992077369
version__ 
owner__davep davep
cc__davep cc__itamarst cc__jknight cc__grib
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
This patch adds a new API to ADBAPI for cases when database
Transactions need to outlive a single interaction.  The use
case that prompted it is a Woven app I am converting to
adbapi.  In it, SQL commit/rollback semantics are used to
ensure that a batch of edits are published together.  An
open SQL transaction can persist across many HTTP requests,
for example. 

The changes are pretty straightforward, with the exception
of the tricky interactions between the threadpool and the
connections used by the ADBAPI ConnectionPool.  Given the
limits of the threadpool API, I think I made just about the
minimal changes, but it looks a little kludgy to me (the bit
with the Semaphore particularly).

Whaddaya think?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Nice, I was going to write that; now I don't have to!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
btw that patch was more like a sketch (worked to initial tests but incomplete).
 I have a better patch with testcases coming up, just wanted to see if the idea
would float or sink.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Hello Mr. Enterprise Maintainer!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
Here's an updated patch (which works, unlike the last one).

dave_p, you there? 

b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Hey Bill, I'm here. I was away from my computer and couldn't
remember my password. This sounds like a cool patch, I'll get
back to you this weekend.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
A couple of comments:

1. The two new _run*InTransaction define a new variable 'conn'
   as the _connection member of the transaction but then don't
   appear to use that variable later.

2. Why do those same two functions call trans.reopen()? Won't
   that commit or rollback whatever query you submitted previously?

3. Could the startTransaction docs mention the need to use the
   transaction with the run*InTransaction methods to avoid blocking?
   Or am I reading that wrong?

4. I think _startTransaction should be using reactor.callFromThread,
   instead of reactor.callInThread so that the transaction will be
   returned by the main loop.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
Hi dave.

Re your comments: 

1. [AGREE] Those are edit-turds.  Sorry. 

2. [DISAGREE] trans.reopen just closes the DBAPI-level cursor object and starts
a new one.  You don't necessarily need a fresh cursor for each query, but it
doesn't hurt.  You can close and open cursors as often as you want. The only
thing to remember is that cursors get invalidated when the transaction they were
created within is committed or rolled-back, so it's best not to count on them
hanging around unless you're real careful.

3. [AGREE] It should mention that you need to commitTransaction or
rollbackTransaction to prevent tying up a thread, for sure.   In general, if
you're going to keep transactions open for decent periods of time, you need to
bump up the cp_max.

4. [DUNNO] If you say so :) 

Also:  My patch accidentally deleted line 239 from adbapi.py, which breaks usage
of the non-Transactified runQuery()... that's the very last line of the patch. 
Ignore that line :)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
There are still some big problems with this approach.  It's very easy to
auto-DOS yourself if you issue more startTransaction calls than cp_max.  I need
to think about how to address this.  Don't commit the current patch :( 

b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
I wasn't planning on committing until the unit tests were ready.

As far as the DOS problem goes, is there really any way to avoid that?
It seems like a built-in consequence of any solution to the problem you
are trying to solve, i.e., if you can reserve transactions for long-term
use, you can reserve too many, too :)

Re my comment #4, I'm pretty sure that callInThread and callFromThread
are paired, with the latter being used to bring out the results of whatever
you started with callInThread. Itamar or someone else will know better, though.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
The DOS problem is avoidable; I have a fix for it.  The problem wasn't that you
could block, it was that you could actually deadlock with no way out... all the
threads tied up in startTransaction and no threads to actually run SQL on those
transactions, including commit() or rollback().  

I've fixed that in my tree by calling the run*InTransaction functions,
commitTransaction, and rollbackTransaction from a separate pool of threads which
have no db connections.

About #4, there's something fishy going on.  My current demo/example app works
if it's callInThread but hangs if callFromThread.  However there are probably
other things going on.  I'll get back to you on that. 

b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Ok, I see what you're saying. And now that I understand it better,
instead of adding another threadpool, you should probably just use
the reactor's threadpool for simplicity. The threadpool in adbapi
is really just for managing the db connections.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Or, hmm. Maybe not. What determines how many transactions
you can have open at once? The size of the new pool?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
The number of db connections is determined by the size of the threadpool; each
thread gets one connection. Transactions are just like any other database
operation, so right now the max number of transactions is cp_max, the max db
connections. 

The threadpool uses next-available scheduling and a single work queue, so when
you submit work to the threadpool you can't know which thread (therefore which
connection) will do the work.  That's why the transaction stuff has to block the
thread that opens the transaction -- to keep it out of the normal work rotation. 

Using the reactor's threadpool won't work with the current threadpool
implementation... there's no way to "reserve" threads to run queries in open
transactions, and even if you could, you'd be wasting available connections
(with cp_max=2, for example, if you opened a transaction, you'd have to reserve
the other thread to run queries in the transaction, which would cause any other
use of the connectionpool to block even though there's a completely unused
connection).

A way to do it without an extra threadpool would be to make per-thread work
queues for threads with open transactions, but &lt;shrug> my current patch
(attached) is IMO simpler than putting all that in place, and doesn't change any
of the user-visible semantics of what cp_min/cp_max mean. 

b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
...but of course there are still problems.  With a pool of threads to run
queries in open transactions, you can get multiple queries on one connection
without proper ordering.  Confusion ensues. 

Still chewing on it,
b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Maybe you could uses Queues instead of Semaphores?
And instead of having a separate threadpool for doing
the work, you could just submit queries to the right queue?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grib's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grib commented|
|-|-|

```
#!html
<pre>
yeah, that's what i've been working on today. I'm pretty close (works most of
the time :)

b.g.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_18"></a>|@jdavisp3 set status to `closed`|
|-|-|

I don't think this patch ever reached a totally working state, and as far as I
know, grib is not in any position to move it there. Persistent transactions might
be useful, but also quite complex, so starting over is probably for the best.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|GitHub commented|
|-|-|

In changeset ec71728efdd9b37775458d668a918067f39fb441
```
#!CommitTicketReference repository="" revision="ec71728efdd9b37775458d668a918067f39fb441"
Merge pull request #303 from twisted/8577-rodrigc-trial-win32-py3

Author: rodrigc
Reviewer: Lukasa
Fixes: #8577

Fix trial on Windows with Python 3
```
